contrib/rpmbuild/src/python/pants/contrib/rpmbuild/tasks/remote_rpm_source_task.py

```diff
@@ -0,0 +1,78 @@
+# coding=utf-8
+# Copyright 2017 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pants.task.task import Task
+from pants.util.dirutil import relative_symlink, safe_mkdir, safe_rmtree
+
+from pants.contrib.rpmbuild.subsystems.remote_rpm_source_fetcher import RemoteRpmSourceFetcher
+from pants.contrib.rpmbuild.targets.remote_rpm_source import RemoteRpmSource
+
+
+class RemoteRpmSourceTask(Task):
+  """Fetch versioned files from a URL.
+
+  Each target will have a stable symlink path suitable for runtime consumption later in the graph.
+  """
+
+  @staticmethod
+  def _is_remote(target):
+    return isinstance(target, RemoteRpmSource)
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(RemoteRpmSourceTask, cls).subsystem_dependencies() + (RemoteRpmSourceFetcher.Factory.scoped(cls),)
+
+  @classmethod
+  def product_types(cls):
+    return ['remote_files', 'runtime_classpath']
+
+  @property
+  def create_target_dirs(self):
+    # Creates the results_dirs but will not cache them.
+    return True
+
+  @property
+  def cache_target_dirs(self):
+    return False
+
+  def stable_root(self, stable_outpath):
+    return os.path.join(self.workdir, 'syms', stable_outpath)
+
+  # TODO(mateo): This works well, but the simplicity is totally obscured by all the case-driven logic around
+  # determining if the returned paths are files or dirs. Make RemoteRpmSourceFetcher return a (dirs, files) abstraction
+  # to clear all that up.
+  def execute(self):
+    targets = self.context.targets(self._is_remote)
+    with self.invalidated(targets, invalidate_dependents=True, topological_order=True) as invalidation_check:
+      # The fetches are idempotent operations from the subsystem, invalidation only controls recreating the symlinks.
+      for vt in invalidation_check.all_vts:
+        remote_rpm_source = RemoteRpmSourceFetcher.Factory.scoped_instance(self).create(vt.target)
+        fetched = remote_rpm_source.path
+        safe_mkdir(fetched)
+
+        # Some unfortunate rigamorole to cover for the case where different targets want the same fetched file
```

Comment:
This is pretty concerning. Why would each target not get its own copy of the fetched package? I don't think the separation of the subsystem that downloads to a private location from the task is a good idea... binary_utils might not be a good thing to emulate when implementing a task.

There is another chunk of code similar to what is happening here that would be really good to emulate/incorporate instead: the RemoteSources and UnpackedJars targets (and relevant tasks) define product-centric dependencies on one another. And in fact, the `UnpackJars` task actually exposes an `unpacked_archives` product (ie: the product is not jar specific).

I don't think it's necessary to refactor away from binary_utils before landing, but one concrete thing that would be good would be to change the `rpm_package` target definition to work with the existing `remote_sources` target. What I think that would look like would be to give `rpm_package` a `sources` argument, and then (for the remote sources case), doing:
```
unpacked_jars(
  name='tarball'
  # TODO: support tarballs.
)

golang_rev = '1.9.2'

remote_sources(
  name='golang',
  sources_target=':tarball',
  dest=rpm_package,
  args={
    spec='golang.spec',
    defines={
      'version': golang_rev,
    },
  },
)
```

You can see an existing example here: https://github.com/pantsbuild/pants/blob/9e5c5ac781278f4075a4a71df03379d72d7aecec/examples/src/protobuf/org/pantsbuild/example/unpacked_jars/BUILD